### ISSUE TITLE
Parser fails on certain files

### ISSUE THREAD

**Issue by ronderksen** [stryker-mutator/stryker#612]
### Summary

I was trying to implement Stryker for one of our repos to see if it was easy to do. Sadly, I've had no luck so far. I ran into two issues: the first one is that the parser crashes on certain files. But the error stacktrace is not helpful (for an end user), because it starts and ends within the node modules.

This is the Stryker output for such an error:
```
$ stryker run
[2018-02-07 08:46:56.360] [INFO] ConfigReader - Using stryker.conf.js in the current working directory.
[2018-02-07 08:46:57.104] [INFO] InputFileResolver - Found 1 of 161 file(s) to be mutated.
[2018-02-07 08:46:57.105] [INFO] InitialTestExecutor - Starting initial test run. This may take a while.
[2018-02-07 08:46:57.106] [INFO] BabelConfigReader - Reading .babelrc file from path "/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/.babelrc"
[2018-02-07 08:47:13.505] [INFO] InitialTestExecutor - Initial test run succeeded. Ran 118 tests in 16 seconds.
[2018-02-07 08:47:13.509] [ERROR] StrykerCli - an error occurred { SyntaxError: Unexpected token (37:6)
    at Parser.pp$5.raise (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:4454:13)
    at Parser.pp.unexpected (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:1761:8)
    at Parser.pp$3.parseIdentifier (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:4332:10)
    at Parser.pp$3.parsePropertyName (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:4156:96)
    at Parser.parsePropertyName (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:6229:23)
    at Parser.pp$3.parseObj (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:4045:12)
    at Parser.pp$3.parseExprAtom (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3719:19)
    at Parser.parseExprAtom (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:7238:22)
    at Parser.pp$3.parseExprSubscripts (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3494:19)
    at Parser.pp$3.parseMaybeUnary (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3474:19)
    at Parser.pp$3.parseExprOps (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3404:19)
    at Parser.pp$3.parseMaybeConditional (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3381:19)
    at Parser.pp$3.parseMaybeAssign (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3344:19)
    at Parser.parseMaybeAssign (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:6474:20)
    at Parser.pp$3.parseParenAndDistinguishExpression (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3828:26)
    at Parser.pp$3.parseExprAtom (/Users/ronderksen/Sites/blaze-wishlist-site-wehkamp/node_modules/babylon/lib/index.js:3709:19) pos: 1631, loc: Position { line: 37, column: 6 } }
```

### Stryker config

```js
const path = require('path');

module.exports = function(config) {
  config.set({
    files: [
      {
        pattern: 'src/app/path/to/file.js',
        mutated: true,
        included: false,
      },
      'src/public/**/*',
      'src/app/**/*.scss',
      'src/app/**/path/to/file.spec.js',
      'src/app/**/path/to/file.spec.js.snap',
      'src/**/*',
      'test/*',
      'development/*',
    ],
    testRunner: 'jest',
    mutator: 'javascript',
    transpilers: ['babel'],
    reporter: ['html', 'clear-text', 'progress'],
    coverageAnalysis: 'off',
    babelrcFile: '.babelrc',
    jest: {
      config: require(path.resolve(__dirname, './test/jest.config.json')),
    },
  });
};

```

### Stryker environment

```
├─┬ stryker@0.18.2
├─┬ stryker-api@0.12.0
├─┬ stryker-babel-transpiler@0.2.4
├─┬ stryker-html-reporter@0.11.5
├─┬ stryker-javascript-mutator@0.3.3
├─┬ stryker-jest-runner@0.4.0
├─┬ stryker-webpack-transpiler@0.1.2
```

### Your Environment
| software         | version(s)
| ---------------- | -------  
| node             |  v8.9.4
| npm              |  v.5.5.1
| Operating System |  macOS 10.13.2

**Comment by nicojs** [comment]
@ronderksen thanks for reporting this issue. Could you share with us the token that the parser is crashing on? It should be in file src/app/path/to/file.js on line 37, column 6 (or line 38, depending on if it counts lines 0-based).

**Comment by ronderksen** [comment]
It looks like it's the spread operator (`...`).

**Comment by ronderksen** [comment]
I just noticed this in the trace logs:
```
[2018-02-07 09:07:12.861] [TRACE] BabelConfigReader - babel config is: undefined
```
Even though the path is correct (pointing to the `.babelrc` in the project's root directory).

**Comment by simondel** [comment]
@ronderksen Awesome, we'll test the spread operator and see what happens :) 

About the babel config trace logging... 
https://github.com/stryker-mutator/stryker/blob/2cda4908564d99db17aea9af186e4ec87fea05ab/packages/stryker-babel-transpiler/src/BabelConfigReader.ts#L10-L14

We put the babelrc file in a variable but we don't actually log the babelrc file, we log the stryker config.. I'll fix that today!

**Comment by nicojs** [comment]
@simondel 
> I'll fix that today!

Too late my friend: https://github.com/stryker-mutator/stryker/pull/614 fixes that

@ronderksen I've tested the spread operator in isolation but it does seem to work with our parser:

```js
function max(...values) {
    Math.max(...values);
}
```

Maybe this example is not representative of what your code looks like? You can pass code through our parser yourself using this piece of code:

```js
const JavaScriptMutator = require('stryker-javascript-mutator/src/JavaScriptMutator').default;
const { Config } = require('stryker-api/config');

const sut = new JavaScriptMutator(new Config());
const files = [
  {
    name: 'testFile.js',
    mutated: true,
    kind: 1,
    content: `
          function max(...values) {
            Math.max(...values);
          }
        `
  }
];
const mutants = sut.mutate(files);
console.log(`mutants: ${JSON.stringify(mutants, null, 2)}`);
```

We're using babylon underneath so whatever it is, we should be able to make it work.

**Comment by ronderksen** [comment]
@nicojs This triggers the error:
```js
const JavaScriptMutator = require('stryker-javascript-mutator/src/JavaScriptMutator').default;
const { Config } = require('stryker-api/config');

const sut = new JavaScriptMutator(new Config());
const files = [
  {
    name: 'testfile.js',
    mutated: true,
    kind: 1,
    content: `
      function test(input) {
        const value = {
          ...input,
          foo: true,
      }
    `
  }
];
const mutants = sut.mutate(files);
console.log(`mutants: ${JSON.stringify(mutants, null, 2)}`);
```

**Comment by nicojs** [comment]
@ronderksen thanks for the clear reproduction steps. It seems that we don't support the `objectRestSpread` yet. See list of plugins here: https://github.com/babel/babel/tree/master/packages/babylon#plugins

I will add it (and some more stage 1, 2 and 3 features) to the default list of plugins. At a later time we should make this part configurable.

**Comment by automeow** [comment]
Hi! I am getting this error. If you log the `this.input` in babylon you can see the file contents it is complaining about. For me it is complaining about this line:
```static propTypes = {```

This passes on the initial run but fails on the mutant runs. The project compiles fine using webpack. There seems to be some issue with mutations and certain es6 flavours?

Cheers.

**Comment by simondel** [comment]
@automeow Thanks for notifying us! I'll test it tomorrow. Just as a check, are you using v0.4.0 of the stryker-javascript-mutator? And could you post your Babel plugins and config?

**Comment by automeow** [comment]
Hi, I've pushed an example repo at https://github.com/automeow/stryker-es6-debug. If you clone it you can see my error by running `npm i && npm run test:mutate`

I'm using 0.3.5, this is what was installed when I did `npm i --save-dev stryker-javascript-mutator`. I didn't realise I should be on 0.4.0. On 0.4.0 my example repo and my real code base both work, so I will use this. Thank you! 😄

**Comment by simondel** [comment]
Glad to hear that the problem has been solved! We usually reply to issues like these with a message which version of which plugin fixed the issue.

So you should have seen something like "This has been fixed with stryker-javascript-mutator 0.4.0" but we forgot that this time :(

Thanks for creating a reproduction repo! Those really help us a lot :) If you need any more help, don't hesitate to contact us!
